driver-core/src/test/java/com/datastax/driver/core/HostConnectionPoolTest.java

```diff
@@ -25,6 +26,130 @@
+            timedOut = true;
+        }
+        assertThat(timedOut).isTrue();
+    }
+
+    @Test(groups = "short")
+    public void should_add_extra_connection_when_core_full() throws ConnectionException, TimeoutException, InterruptedException {
+        cluster.getConfiguration().getPoolingOptions()
+            .setIdleTimeoutSeconds(20)
+            .setMinSimultaneousRequestsPerConnectionThreshold(HostDistance.LOCAL, 1)
+            .setMaxConnectionsPerHost(HostDistance.LOCAL, 128);
+
+        HostConnectionPool pool = createPool(1, 2);
+        PooledConnection core = pool.connections.get(0);
+
+        // Fill core connection
+        for (int i = 0; i < 128; i++)
+            assertThat(
+                pool.borrowConnection(100, MILLISECONDS)
+            ).isEqualTo(core);
+
+        // Reaching 128 on the core connection should have triggered the creation of an extra one
+        TimeUnit.MILLISECONDS.sleep(100);
+        assertThat(pool.connections).hasSize(2);
+        PooledConnection extra1 = pool.connections.get(1);
+
+        assertThat(
+            pool.borrowConnection(100, MILLISECONDS)
+        ).isEqualTo(extra1);
+
+        // If the extra connection is returned it gets trashed
+        pool.returnConnection(extra1);
+        assertThat(pool.connections).hasSize(1);
+
+        // If the core connection gets returned we can borrow it again
+        pool.returnConnection(core);
+        assertThat(
+            pool.borrowConnection(100, MILLISECONDS)
+        ).isEqualTo(core);
+    }
+
+    @Test(groups = "short")
+    public void should_resurrect_trashed_connection_within_idle_timeout() throws ConnectionException, TimeoutException, InterruptedException {
+        cluster.getConfiguration().getPoolingOptions()
+            .setIdleTimeoutSeconds(20)
+            .setMinSimultaneousRequestsPerConnectionThreshold(HostDistance.LOCAL, 1)
+            .setMaxConnectionsPerHost(HostDistance.LOCAL, 128);
+
+        HostConnectionPool pool = createPool(1, 2);
+        PooledConnection core = pool.connections.get(0);
+
+        for (int i = 0; i < 128; i++)
+            assertThat(pool.borrowConnection(100, MILLISECONDS)).isEqualTo(core);
+
+        TimeUnit.MILLISECONDS.sleep(100);
+        assertThat(pool.connections).hasSize(2);
+        PooledConnection extra1 = pool.connections.get(1);
+
+        assertThat(pool.borrowConnection(100, MILLISECONDS)).isEqualTo(extra1);
+        pool.returnConnection(extra1);
+        assertThat(pool.connections).hasSize(1);
+
+        // extra1 is now in the trash, core still full
+        // Borrowing again should resurrect extra1 from the trash
+        assertThat(
+            pool.borrowConnection(100, MILLISECONDS)
+        ).isEqualTo(extra1);
+        assertThat(pool.connections).hasSize(2);
+    }
+
+    @Test(groups = "long")
+    public void should_not_resurrect_trashed_connection_after_idle_timeout() throws ConnectionException, TimeoutException, InterruptedException {
```

Comment:
should_not_resurrect_trashed_connection_after_idle_timeout and should_resurrect_trashed_connection_within_idle_timeout are almost equivalent except for the last piece of the validation.  Might be nice to create a separate utility method that creates a non-core connection, validates it's use and returns it.   Though I can see the argument for keeping the test cases separate and fully defined for understanding.